enhancements/network/ovn-secondary-networks.md

```diff
@@ -0,0 +1,332 @@
+- these overlays can optionally be a subnet
+- IPv6 support is welcome (not an immediate priority)
+- optional connection to the external networks
+
+### Non-Goals
+- network policy concerns. There isn't a requirement for having network
+  policies for the secondary networks. There is an
+  [enhancement proposal](enhancements/network/multi-networkpolicy.md) about
+  this subject, not scoped around OVN.
+- ingress into the OVN secondary networks. This is not a current objective,
+  and should be treated in a separate enhancement proposal, which would depend
+  on the [services on secondary networks feature](TODO: add link to PR) availability.
+
+## Proposal
+
+### User Stories
+
+#### Isolated tenant network
+A customer can specify a layer 2 overlay network with a subnet via a
+`network-attachment-definition`, which is provisioned without the intervention
+of a system administrator. You can refer to the
+[network attachment definition section](#net-attach-def-samples) for examples.
+
+Afterwards, the users can then request their pods - and / VMs - to connect to
+those overlays as they would for any other secondary network.
+
+### Implementation Details/Notes/Constraints
+A new CRD will be introduced, named `OverlayNetwork`. When this CRD is
+provisioned, a `NetworkAttachmentDefinition` object will be rendered.
+
+The new CRD will look like:
+
+```golang
+type OverlayNetwork struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec OverlayNetworkSpec `json:"spec,omitempty"`
+}
+
+type OverlayNetworkSpec struct {
+    // Subnet is a RFC 4632/4291-style string that represents an IP address and prefix length in CIDR notation
+    Subnet                  string `json:"subnet"`
+    HasExternalConnectivity bool   `json:"hasExternalConnectivity,omitempty"`
+    MTU                     uint16 `json:"mtu,omitempty"`
+}
+```
+
+A controller will watch out for the creation / deletion of these
+`OverlayNetwork`s, and will render and provision the corresponding
+`NetworkAttachmentDefinition`s.
+
+**Note**: the subnet attribute requires some validation, especially *if* the
+subnet has external connectivity: we need to assure the uniqueness of the
+subnet on these overlay networks, and when the provisioned overlay has external
+connectivity, we need to furthermore assure that its range does not clash with
+the node CIDRs, cluster CIDR, and service CIDR. An error will be thrown when
+the provided range clashes with the aforementioned CIDRs.
+
+#### External connectivity considerations
+When an overlay network with external connectivity is requested, the traffic
+will be directly dumped onto the host's underlay via a localnet port.
+
+#### Network-attachment-definition provisioning
+A removed `Network-Attachment-definition` with present attachment will
+trigger a reconciliation loop, which will cause the net-attach-def entity to be
+reprovisioned; if the network attachment definition is removed and the
```

Comment:
This seems backwards; deleting the OverlayNetwork should cause the switch to be deleted.

Honestly, I wouldn't even wait for the switch to be empty to delete it; it is always possible to miss CNI DEL calls.

Reply:
But what if the logical switch is in use ? i.e. pods are up, connected to the logical switch, doing whatever it is they do. 

this would remove the logical switch, along w/ the LSPs connected to it, breaking connectivity among those pods.

Reply:
Then don't delete the network?

It's not good form to keep artifacts around if the representative Kubernetes object is missing. What happens if the object is re-created with the same name while the switch is still extant, but with different IPs? Will those pods suddenly have half-broken connectivity?

So, either the artifacts are deleted as soon as the object is deleted, or you set a Finalizer that keeps the object in Terminating state until something is cleaned up. Finalizers are tricky to get right, so I'd recommend just deleting the underlying switch.

Reply:
> What happens if the object is re-created with the same name while the switch is still extant, but with different IPs?

That said, we need to handle this case as part of being level-triggered. We could easily miss a delete-add pair. The joys of synchronizing two data sources.